### PR TITLE
Adding Electron Support

### LINKED ISSUE #356: Add electron as a target?
**Enhancement** 

From the small playing around with electron (at: https://github.com/matt-gadd/todo-mvc-electron). It seems like it'd be fairly trivial to add basic electron support as a target for cli-build-app. Some random thoughts:

* In dev we can support live reload of the `BrowserWindow` in electron by pointing to the dojo served app. This would be a little more work for the main process code.

* At the moment the example above just uses `ts-node` for the main process code, we might want to deal with that in build-app, or actually bundle it properly in webpack too.

* For a dist output we should be able to use https://github.com/electron/electron-packager, to actually ship a binary.

* Perhaps we can make the `BrowserWindow` options declarative in the `.dojorc`, and internal-ize the basic electron boilerplate (https://github.com/matt-gadd/todo-mvc-electron/blob/master/src/electron.ts). This would mean for a basic app (that uses no electron features), no code would be required at all.

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/)
* [x] Unit or Functional tests are included in the PR
* [x] schema.json has been updated appopriately

**Description:**

Adding Electron target. This is largely based on https://github.com/matt-gadd/cli-build-app/tree/electron-support.

Resolves #356

### CODE REVIEW

File: src/electron.config.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		},
+		mode: args.mode === 'dev' ? 'development' : 'production',
+		devtool: 'source-map',
+		watchOptions: { ignored: /node_modules/ },
+		target: 'electron-main',
+		plugins: [
+			new ElectronPlugin({
+				electron: {
+					browser: electron.browser || {},
+					packaging: electron.packaging || {}
```
Comment (rorticus) [inline src/electron.config.ts:45]: These `packaging` options don't do anything interesting until the new `ElectronPlugin` is released.

File: src/main.ts [hunk 2]
```diff
@@ -37,20 +37,19 @@ function getLibraryName(name: string) {
...
 
 const testModes = ['test', 'unit', 'functional'];
 
-function createCompiler(config: webpack.Configuration) {
-	const compiler = webpack(config);
-	fixMultipleWatchTrigger(compiler);
-	return compiler;
+// for some reason the MultiCompiler type doesn't include hooks, even though they are clearly defined on the
+// object coming back.
+interface MultiCompilerWithHooks extends webpack.MultiCompiler {
```
Comment (rorticus) [inline src/main.ts:47]: Is there a better way to do this?

File: src/main.ts [hunk 3]
```diff
@@ -162,19 +167,21 @@ async function serve(config: webpack.Configuration, args: any) {
...
 		}
-		const jsonpName = (config.output && config.output.jsonpFunction) || 'unknown';
+		const jsonpName = (mainConfig.output && mainConfig.output.jsonpFunction) || 'unknown';
 		const onDemandBtr = new OnDemandBtr({
 			buildTimeRenderOptions: btrOptions,
 			scope: libraryName,
 			base,
-			compiler,
-			entries: config.entry ? Object.keys(config.entry) : [],
+			compiler: compiler.compilers[0],
```
Comment (rorticus) [inline src/main.ts:183]: No need to pass in electron compiler to BTR. Realistically, probably no reason to pass _anything_ to BTR in this case...

File: src/main.ts [hunk 4]
```diff
@@ -355,35 +369,40 @@ const command: Command = {
...
 		} else if (args.mode === 'functional') {
-			config = functionalConfigFactory(args);
+			configs.push(functionalConfigFactory(args));
 		} else {
-			config = distConfigFactory(args);
+			configs.push(distConfigFactory(args));
+		}
+
+		if (args.target === 'electron') {
+			args.base = `.${args.base}`;
```
Comment (rorticus) [inline src/main.ts:390]: If we're using electron, the base path can't be absolute and must be relative.

File: package.json [hunk 5]
```diff
@@ -119,6 +119,7 @@
     "css-loader": "1.0.1",
     "css-url-relative-plugin": "1.0.0",
     "cssnano": "4.1.7",
+    "electron": "10.1.1",
```
Comment (rorticus) [inline package.json:122]: Adding `electron` as a dep here so you dont need to install it separately on every app. In the future, we'll look into how to conditionally install deps like puppetteer and electron.

File: src/dev.config.ts [hunk 6]
```diff
@@ -24,7 +24,7 @@ function webpackConfig(args: any): webpack.Configuration {
 	const experimental = args.experimental || {};
 	const isExperimentalSpeed = !!experimental.speed;
 	const singleBundle = args.singleBundle || isExperimentalSpeed;
-	const base = args.base || '/';
+	const base = args.target === 'electron' ? './' : args.base || '/';
```
Comment (rorticus) [inline src/dev.config.ts:27]: Electron target does not support a base URL

File: src/dist.config.ts [hunk 7]
```diff
@@ -33,7 +33,7 @@ All rights reserved
 
 function webpackConfig(args: any): webpack.Configuration {
 	const basePath = process.cwd();
-	const base = args.base || '/';
+	const base = args.target === 'electron' ? './' : args.base || '/';
```
Comment (rorticus) [inline src/dist.config.ts:36]: Electron target does not support a base URL

File: src/electron.config.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+				},
+				watch: !!args.watch,
+				serve: !!args.serve,
+				port: args.port,
+				dist: args.mode !== 'dev'
+			} as any)
```
Comment (rorticus) [inline src/electron.config.ts:51]: Casting this as `any` until new ElectronPlugin is released

**Review by matt-gadd — APPROVED** [review]

**Pull request merged**